src/deltalake_ext.rs

```diff
@@ -0,0 +1,322 @@
+    DeltaTable { 
+        #[from]
+        source: DeltaTableError 
+    },
+
+    #[error("Arrow interaction failed: {source}")]
+    Arrow { 
+        #[from]
+        source: ArrowError 
+    },
+
+    #[error("Parquet write failed: {source}")]
+    Parquet { 
+        #[from]
+        source: ParquetError 
+    },
+}
+
+pub struct DeltaParquetWriter {
+    table_path: String,
+    storage: Box<dyn StorageBackend>,
+    arrow_schema_ref: Arc<arrow::datatypes::Schema>,
+    writer_properties: WriterProperties,
+    cursor: InMemoryWriteableCursor,
+    arrow_writer: ArrowWriter<InMemoryWriteableCursor>,
+    partition_values: HashMap<String, String>,
+    num_records: i64,
+}
+
+/// A writer that writes record batches in parquet format to a table location.
+/// This should be used along side a DeltaTransaction wrapping the same DeltaTable instance.
+impl DeltaParquetWriter {
+    /// Initialize the writer from the given table path and delta schema
+    pub async fn for_table_path_with_schema(table_path: String, schema: &Schema) -> Result<DeltaParquetWriter, DeltaWriterError> {
+        // Initialize storage backend for the given table path
+        let storage = deltalake::get_backend_for_uri(table_path.as_str())?;
+
+        // Initialize an arrow schema ref from the delta table schema
+        let arrow_schema = <ArrowSchema as From<&Schema>>::from(schema);
+        let arrow_schema_ref = Arc::new(arrow_schema);
+
+        // Initialize writer properties for the underlying arrow writer
+        let writer_properties = WriterProperties::builder()
+            // TODO: Extract config/env for writer properties and set more than just compression
+            .set_compression(Compression::SNAPPY)
+            .build();
+
+        let cursor = InMemoryWriteableCursor::default();
+        // TODO: Handle error
+        let arrow_writer = ArrowWriter::try_new(cursor.clone(), arrow_schema_ref.clone(), Some(writer_properties.clone())).unwrap();
+
+        Ok(Self {
+            table_path,
+            storage,
+            arrow_schema_ref,
+            writer_properties,
+            cursor,
+            arrow_writer,
+            partition_values: HashMap::new(),
+            num_records: 0,
+        })
+    }
+
+    /// Writes the record batch in-memory and updates internal state accordingly.
+    /// This method buffers the write stream internally so it can be invoked for many record batches and flushed after the appropriate number of bytes has been written.
+    pub async fn write_record_batch(&mut self, partition_cols: &Vec<String>, record_batch: &RecordBatch) -> Result<(), DeltaWriterError> {
+        let partition_values = extract_partition_values(partition_cols, record_batch)?;
+
+        // TODO: Verify that this RecordBatch contains partition values that agree with previous partition values (if any) for the current writer context
+        self.partition_values = partition_values;
+
+        // TODO: collect/update column stats
```

Comment:
Oh fantastic idea! That would simplify matters and get rid of the need for more passes over the same batch.🤞

Reply:
For now - I've just added a comment in the code referring to this conversation in https://github.com/delta-io/kafka-delta-ingest/pull/1/files#diff-5d8c9545386ed0b9dcf9dab1722a6b484e1d3014d9a2470788dd7b20509ca642R147. Will wait to hear from you on the feasibility @nevi-me (again - great recommendation).

Reply:
I've been thinking about this more @xianwill, I think it might be more fruitful for us to initially calculate the stats using the arrow compute facility.
I'm still a bit uncertain that we can share the stats with the just-written parquet file, as the file's statistic granularity is at the page + column level, while the entire file is only at a column level.

From the protocol format:

> Global statistics record information about the entire file.

Perhaps we can bootstrap this after this PR is merged.

Reply:
Yes - I think this will be a good follow-up issue to mature the delta writer after initial merge.

Reply:
For some of the stats like max, min and count, will we be able to calculate the global column stats using page segmented column stats produced from the parquet write?